File: backend/services/whitelist_service.py
"""
Whitelist Service - Maintains known legitimate projects and addresses
Prevents false positives by identifying official tokens, exchanges, and verified projects
"""

import asyncio
import json
import logging
from typing import Set, Dict, List, Optional, Any
from datetime import datetime
import aiohttp
from aiohttp import ClientTimeout
import sys
import os
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config.redis_config import redis_config

logger = logging.getLogger(__name__)


class WhitelistService:
    def __init__(self):
        self.redis_client = redis_config.get_client()
        self.legitimate_addresses: Set[str] = set()
        self.verified_tokens: Dict[str, Dict] = {}
        self.official_exchanges: Set[str] = set()
        self.last_update = None
        self.update_interval = 3600  # 1 hour

        # Initialize known legitimate addresses manually
        self._init_static_whitelist()

        # Whitelist sources
        self.sources = {
            'solana_token_list': {
                'url': 'https://raw.githubusercontent.com/solana-labs/token-list/main/src/tokens/solana.tokenlist.json',
                'parser': self._parse_solana_token_list
            },
            'coingecko_verified': {
                'url': 'https://api.coingecko.com/api/v3/coins/list?include_platform=true',
                'parser': self._parse_coingecko_verified
            }
        }

        # Log Redis status
        if redis_config.is_available():
            conn_info = redis_config.get_connection_info()
            logger.info(f"Whitelist using Redis: {conn_info['type']} ({conn_info['status']})")
        else:
            logger.warning("Whitelist running without Redis cache")

    def _init_static_whitelist(self):
        """Initialize manually curated whitelist of known legitimate addresses."""

        # Major tokens on Solana
        self.verified_tokens = {
            # Native SOL
            "So11111111111111111111111111111111111111112": {
                "name": "Wrapped SOL",
                "symbol": "SOL",
                "type": "native",
                "verified": True,
                "official": True
            },
            # USDC
            "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v": {
                "name": "USD Coin",
                "symbol": "USDC",
                "type": "stablecoin",
                "verified": True,
                "official": True
            },
            # USDT
            "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB": {
                "name": "Tether USD",
                "symbol": "USDT",
                "type": "stablecoin",
                "verified": True,
                "official": True
            },
            # SAMO
            "7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU": {
                "name": "Samoyedcoin",
                "symbol": "SAMO",
                "type": "memecoin",
                "verified": True,
                "official": True
            },
            # mSOL
            "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So": {
                "name": "Marinade staked SOL",
                "symbol": "mSOL",
                "type": "liquid_staking",
                "verified": True,
                "official": True
            },
            # RAY
            "4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R": {
                "name": "Raydium",
                "symbol": "RAY",
                "type": "defi",
                "verified": True,
                "official": True
            },
            # SRM
            "SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt": {
                "name": "Serum",
                "symbol": "SRM",
                "type": "defi",
                "verified": True,
                "official": True
            }
        }

        # Major exchanges and official programs
        self.official_exchanges = {
            # Raydium
            "675kPX9MHTjS2zt1qfr1NYHuzeLXfQM9H24wFSUt1Mp8",
            "5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1",
            # Orca
            "whirLbMiicVdio4qvUfM5KAg6Ct8VwpYzGff3uctyCc",
            "9W959DqEETiGZocYWCQPaJ6sBmUzgfxXfqGeTEdp3aQP",
            # Jupiter
            "JUP4Fb2cqiRUcaTHdrPC8h2gNsA2ETXiPDD33WcGuJB",
            "JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4",
            # Serum
            "9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin",
            "EhqXDsotvYZGAzGJKS4t1JthyaS6U6mE2JNGNwFYkrn",
            # System programs
            "11111111111111111111111111111111",  # System Program
            "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",  # Token Program
        }

    async def check_address_legitimacy(self, address: str) -> Dict[str, Any]:
        """
        Check if an address is known to be legitimate.
        Returns comprehensive legitimacy information.
        """
        try:
            # Check cache first
            cached_result = await self._get_cached_legitimacy(address)
            if cached_result:
                return cached_result

            # Check static whitelist
            legitimacy_info = {
                "address": address,
                "is_legitimate": False,
                "confidence": 0.0,
                "sources": [],
                "token_info": None,
                "exchange_info": None,
                "verification_level": "unknown",
                "last_checked": datetime.now().isoformat()
            }

            # Check if it's a verified token
            if address in self.verified_tokens:
                token_info = self.verified_tokens[address]
                legitimacy_info.update({
                    "is_legitimate": True,
                    "confidence": 0.95,
                    "sources": ["static_whitelist", "verified_tokens"],
                    "token_info": token_info,
                    "verification_level": "official" if token_info.get("official") else "verified"
                })

            # Check if it's an official exchange/program
            elif address in self.official_exchanges:
                legitimacy_info.update({
                    "is_legitimate": True,
                    "confidence": 0.90,
                    "sources": ["static_whitelist", "official_exchanges"],
                    "exchange_info": {"type": "official_program"},
                    "verification_level": "official"
                })

            # If not in static lists, check dynamic sources
            else:
                await self._check_dynamic_sources(address, legitimacy_info)

            # Cache the result
            await self._cache_legitimacy_result(address, legitimacy_info)

            return legitimacy_info

        except Exception as e:
            logger.error(f" Error checking address legitimacy: {e}")
            return {
                "address": address,
                "is_legitimate": False,
                "confidence": 0.0,
                "sources": [],
                "error": str(e),
                "verification_level": "unknown",
                "last_checked": datetime.now().isoformat()
            }

    async def _check_dynamic_sources(self, address: str, legitimacy_info: Dict):
        """Check dynamic sources for address legitimacy."""

        # Check if we need to update from external sources
        if await self._should_update_whitelist():
            await self._update_from_sources()

        # Check against updated token list
        if address in self.verified_tokens:
            token_info = self.verified_tokens[address]
            legitimacy_info.update({
                "is_legitimate": True,
                "confidence": 0.85,
                "sources": legitimacy_info["sources"] + ["dynamic_token_list"],
                "token_info": token_info,
                "verification_level": "verified"
            })

    async def _should_update_whitelist(self) -> bool:
        """Check if whitelist needs updating."""
        if not self.last_update:
            return True

        time_since_update = datetime.now() - self.last_update
        return time_since_update.total_seconds() > self.update_interval

    async def _update_from_sources(self):
        """Update whitelist from external sources."""
        try:
            logger.info(" Updating whitelist from external sources...")

            for source_name, source_config in self.sources.items():
                try:
                    await self._fetch_and_parse_source(source_name, source_config)
                except Exception as e:
                    logger.warning(f" Failed to update from {source_name}: {e}")

            self.last_update = datetime.now()
            logger.info(f" Whitelist updated. Total verified tokens: {len(self.verified_tokens)}")

        except Exception as e:
            logger.error(f" Error updating whitelist: {e}")

    async def _fetch_and_parse_source(self, source_name: str, source_config: Dict):
        """Fetch and parse a specific whitelist source."""

        timeout = ClientTimeout(total=10)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(source_config['url']) as response:
                if response.status == 200:
                    data: Any = await response.json()
                    await source_config['parser'](data, source_name)
                else:
                    logger.warning(f" Failed to fetch {source_name}: HTTP {response.status}")

    async def _parse_solana_token_list(self, data: Dict, source_name: str):
        """Parse official Solana token list."""
        try:
            tokens = data.get('tokens', [])
            for token in tokens:
                address = token.get('address')
                if address and address not in self.verified_tokens:
                    self.verified_tokens[address] = {
                        "name": token.get('name', 'Unknown'),
                        "symbol": token.get('symbol', 'UNK'),
                        "type": "token",
                        "verified": True,
                        "official": True,
                        "source": source_name
                    }

            logger.info(f" Parsed {len(tokens)} tokens from {source_name}")

        except Exception as e:
            logger.error(f" Error parsing {source_name}: {e}")

    async def _parse_coingecko_verified(self, data: List, source_name: str):
        """Parse CoinGecko verified tokens."""
        try:
            solana_tokens = []
            for coin in data:
                platforms = coin.get('platforms', {})
                if 'solana' in platforms and platforms['solana']:
                    address = platforms['solana']
                    if address and address not in self.verified_tokens:
                        self.verified_tokens[address] = {
                            "name": coin.get('name', 'Unknown'),
                            "symbol": coin.get('symbol', 'UNK').upper(),
                            "type": "token",
                            "verified": True,
                            "official": False,
                            "source": source_name,
                            "coingecko_id": coin.get('id')
                        }
                        solana_tokens.append(address)

            logger.info(f" Parsed {len(solana_tokens)} Solana tokens from {source_name}")

        except Exception as e:
            logger.error(f" Error parsing {source_name}: {e}")

    async def _get_cached_legitimacy(self, address: str) -> Optional[Dict]:
        """Get cached legitimacy result."""
        if not self.redis_client:
            return None

        try:
            cached = self.redis_client.get(f"whitelist:{address}")
            if cached and isinstance(cached, (str, bytes)):
                return json.loads(cached)
        except Exception as e:
            logger.warning(f" Cache read error: {e}")

        return None

    async def _cache_legitimacy_result(self, address: str, result: Dict):
        """Cache legitimacy result."""
        if not self.redis_client:
            return

        try:
            # Cache for 1 hour
            self.redis_client.setex(
                f"whitelist:{address}",
                3600,
                json.dumps(result, default=str)
            )
        except Exception as e:
            logger.warning(f" Cache write error: {e}")

    async def is_legitimate_project(self, address: str) -> bool:
        """Simple boolean check for legitimacy."""
        result = await self.check_address_legitimacy(address)
        return result.get("is_legitimate", False)

    async def get_legitimacy_confidence(self, address: str) -> float:
        """Get confidence score for legitimacy (0.0 to 1.0)."""
        result = await self.check_address_legitimacy(address)
        return result.get("confidence", 0.0)

    def add_to_whitelist(self, address: str, info: Dict):
        """Manually add address to whitelist."""
        self.verified_tokens[address] = {
            **info,
            "verified": True,
            "source": "manual",
            "added_at": datetime.now().isoformat()
        }
        logger.info(f" Added {address} to whitelist manually")

    def remove_from_whitelist(self, address: str):
        """Remove address from whitelist."""
        if address in self.verified_tokens:
            del self.verified_tokens[address]
            logger.info(f" Removed {address} from whitelist")

    async def get_whitelist_stats(self) -> Dict:
        """Get whitelist statistics."""
        return {
            "total_verified_tokens": len(self.verified_tokens),
            "total_official_exchanges": len(self.official_exchanges),
            "last_update": self.last_update.isoformat() if self.last_update else None,
            "cache_enabled": self.redis_client is not None,
            "sources_configured": len(self.sources)
        }


# Global instance
whitelist_service = WhitelistService()

File: juliaos/core/src/security/WhitelistService.jl
"""
WhitelistService.jl - High-performance legitimate address verification service

Migrated from backend/services/whitelist_service.py for native Julia performance.
Maintains whitelists of verified tokens, exchanges, and official programs to prevent false positives.
"""
module WhitelistService

using HTTP
using JSON3
using Dates
using Logging

export check_address_legitimacy, is_legitimate_project, get_legitimacy_confidence, get_whitelist_stats

# ========================================
# DATA STRUCTURES & STATE
# ========================================

mutable struct WhitelistState
    verified_tokens::Dict{String, Dict{String, Any}}
    official_exchanges::Set{String}
    last_update::Union{DateTime, Nothing}
    update_interval_seconds::Int
    lock::ReentrantLock
end

# Sources for dynamic whitelist updates
const SOURCES = Dict(
    "solana_token_list" => "https://raw.githubusercontent.com/solana-labs/token-list/main/src/tokens/solana.tokenlist.json",
    "coingecko_verified" => "https://api.coingecko.com/api/v3/coins/list?include_platform=true"
)

# Global state
const STATE = WhitelistState(
    Dict{String, Dict{String, Any}}(),
    Set{String}(),
    nothing,
    3600,  # 1 hour
    ReentrantLock()
)

# ========================================
# INITIALIZATION
# ========================================

"""
    initialize_whitelist()

Initialize the whitelist service with static data and fetch dynamic sources.
"""
function initialize_whitelist()
    @info "Initializing WhitelistService..."

    _init_static_whitelist!()
    _init_official_exchanges!()

    # Update from dynamic sources
    try
        update_from_sources!()
        @info "Whitelist initialized successfully with $(length(STATE.verified_tokens)) tokens and $(length(STATE.official_exchanges)) exchanges."
    catch e
        @warn "Failed to update from dynamic sources during initialization" error=e
        @info "Whitelist initialized with static data only: $(length(STATE.verified_tokens)) tokens and $(length(STATE.official_exchanges)) exchanges."
    end
end

"""
Initialize static whitelist with manually curated legitimate addresses
"""
function _init_static_whitelist!()
    lock(STATE.lock) do
        # Major tokens on Solana
        STATE.verified_tokens = Dict{String, Dict{String, Any}}(
            # Native SOL
            "So11111111111111111111111111111111111111112" => Dict(
                "name" => "Wrapped SOL",
                "symbol" => "SOL",
                "type" => "native",
                "verified" => true,
                "official" => true
            ),
            # USDC
            "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v" => Dict(
                "name" => "USD Coin",
                "symbol" => "USDC",
                "type" => "stablecoin",
                "verified" => true,
                "official" => true
            ),
            # USDT
            "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB" => Dict(
                "name" => "Tether USD",
                "symbol" => "USDT",
                "type" => "stablecoin",
                "verified" => true,
                "official" => true
            ),
            # SAMO
            "7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU" => Dict(
                "name" => "Samoyedcoin",
                "symbol" => "SAMO",
                "type" => "memecoin",
                "verified" => true,
                "official" => true
            ),
            # mSOL
            "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So" => Dict(
                "name" => "Marinade staked SOL",
                "symbol" => "mSOL",
                "type" => "liquid_staking",
                "verified" => true,
                "official" => true
            ),
            # RAY
            "4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R" => Dict(
                "name" => "Raydium",
                "symbol" => "RAY",
                "type" => "defi",
                "verified" => true,
                "official" => true
            ),
            # SRM
            "SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt" => Dict(
                "name" => "Serum",
                "symbol" => "SRM",
                "type" => "defi",
                "verified" => true,
                "official" => true
            )
        )
    end
end

"""
Initialize official exchanges and programs
"""
function _init_official_exchanges!()
    lock(STATE.lock) do
        STATE.official_exchanges = Set{String}([
            # Raydium
            "675kPX9MHTjS2zt1qfr1NYHuzeLXfQM9H24wFSUt1Mp8",
            "5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1",
            # Orca
            "whirLbMiicVdio4qvUfM5KAg6Ct8VwpYzGff3uctyCc",
            "9W959DqEETiGZocYWCQPaJ6sBmUzgfxXfqGeTEdp3aQP",
            # Jupiter
            "JUP4Fb2cqiRUcaTHdrPC8h2gNsA2ETXiPDD33WcGuJB",
            "JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4",
            # Serum
            "9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin",
            "EhqXDsotvYZGAzGJKS4t1JthyaS6U6mE2JNGNwFYkrn",
            # System programs
            "11111111111111111111111111111111",  # System Program
            "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",  # Token Program
        ])
    end
end

# ========================================
# MAIN API FUNCTIONS
# ========================================

"""
    check_address_legitimacy(address::String)::Dict

Check if an address is known to be legitimate.
Returns comprehensive legitimacy information with confidence scores.
"""
function check_address_legitimacy(address::String)::Dict
    try
        # Auto-update if needed
        if _should_update()
            @info "Whitelist is stale, triggering update..."
            Threads.@spawn update_from_sources!()
        end

        legitimacy_info = Dict(
            "address" => address,
            "is_legitimate" => false,
            "confidence" => 0.0,
            "sources" => String[],
            "token_info" => nothing,
            "exchange_info" => nothing,
            "verification_level" => "unknown",
            "last_checked" => now(UTC)
        )

        # Check verified tokens
        if haskey(STATE.verified_tokens, address)
            token_info = STATE.verified_tokens[address]
            legitimacy_info["is_legitimate"] = true
            legitimacy_info["confidence"] = token_info["official"] ? 0.95 : 0.85
            legitimacy_info["sources"] = ["static_whitelist", "verified_tokens"]
            legitimacy_info["token_info"] = token_info
            legitimacy_info["verification_level"] = token_info["official"] ? "official" : "verified"

        # Check official exchanges/programs
        elseif address in STATE.official_exchanges
            legitimacy_info["is_legitimate"] = true
            legitimacy_info["confidence"] = 0.90
            legitimacy_info["sources"] = ["static_whitelist", "official_exchanges"]
            legitimacy_info["exchange_info"] = Dict("type" => "official_program")
            legitimacy_info["verification_level"] = "official"
        end

        return legitimacy_info

    catch e
        @error "Error checking address legitimacy" address=address error=e
        return Dict(
            "address" => address,
            "is_legitimate" => false,
            "confidence" => 0.0,
            "sources" => String[],
            "error" => string(e),
            "verification_level" => "unknown",
            "last_checked" => now(UTC)
        )
    end
end

"""
    is_legitimate_project(address::String)::Bool

Simple boolean check for legitimacy.
"""
function is_legitimate_project(address::String)::Bool
    result = check_address_legitimacy(address)
    return get(result, "is_legitimate", false)
end

"""
    get_legitimacy_confidence(address::String)::Float64

Get confidence score for legitimacy (0.0 to 1.0).
"""
function get_legitimacy_confidence(address::String)::Float64
    result = check_address_legitimacy(address)
    return get(result, "confidence", 0.0)
end

"""
    get_whitelist_stats()::Dict

Get whitelist statistics and status information.
"""
function get_whitelist_stats()::Dict
    return Dict(
        "total_verified_tokens" => length(STATE.verified_tokens),
        "total_official_exchanges" => length(STATE.official_exchanges),
        "last_update" => STATE.last_update,
        "cache_type" => "In-Memory",
        "sources_configured" => length(SOURCES),
        "update_interval_minutes" => STATE.update_interval_seconds ÷ 60
    )
end

# ========================================
# DYNAMIC UPDATES
# ========================================

"""
    update_from_sources!()

Update whitelist from external sources in parallel.
"""
function update_from_sources!()
    lock(STATE.lock) do
        @info "Updating whitelist from external sources..."

        tasks = []
        for (source_name, url) in SOURCES
            task = Threads.@spawn _fetch_and_parse_source(source_name, url)
            push!(tasks, task)
        end

        # Wait for all tasks and collect results
        results = fetch.(tasks)

        # Process results
        for (i, result) in enumerate(results)
            source_name = collect(keys(SOURCES))[i]
            if result isa Dict && !isempty(result)
                # Merge new tokens into existing whitelist
                merge!(STATE.verified_tokens, result)
                @info "Updated from $source_name: $(length(result)) tokens"
            else
                @warn "No data received from $source_name"
            end
        end

        STATE.last_update = now(UTC)
        @info "Whitelist update completed. Total verified tokens: $(length(STATE.verified_tokens))"
    end
end

"""
Fetch and parse a single whitelist source
"""
function _fetch_and_parse_source(source_name::String, url::String)::Dict{String, Dict{String, Any}}
    try
        @info "Fetching from $source_name..."
        response = HTTP.get(url, Dict("User-Agent" => "Ghost-Wallet-Hunter-Julia-Client/1.0"))

        if response.status == 200
            data = JSON3.read(response.body)

            if source_name == "solana_token_list"
                return _parse_solana_token_list(data, source_name)
            elseif source_name == "coingecko_verified"
                return _parse_coingecko_verified(data, source_name)
            else
                @warn "No parser for source: $source_name"
            end
        else
            @warn "Failed to fetch $source_name" status=response.status
        end
    catch e
        @error "Error fetching or parsing $source_name" url=url error=e
    end

    return Dict{String, Dict{String, Any}}()
end

"""
Parse official Solana token list
"""
function _parse_solana_token_list(data, source_name::String)::Dict{String, Dict{String, Any}}
    tokens = Dict{String, Dict{String, Any}}()

    try
        if haskey(data, "tokens")
            for token in data.tokens
                if haskey(token, "address") && !haskey(STATE.verified_tokens, token.address)
                    tokens[token.address] = Dict(
                        "name" => get(token, "name", "Unknown"),
                        "symbol" => get(token, "symbol", "UNK"),
                        "type" => "token",
                        "verified" => true,
                        "official" => true,
                        "source" => source_name
                    )
                end
            end
        end

        @info "Parsed $(length(tokens)) tokens from $source_name"
    catch e
        @error "Error parsing $source_name" error=e
    end

    return tokens
end

"""
Parse CoinGecko verified tokens
"""
function _parse_coingecko_verified(data, source_name::String)::Dict{String, Dict{String, Any}}
    tokens = Dict{String, Dict{String, Any}}()

    try
        solana_count = 0
        for coin in data
            if haskey(coin, "platforms") && haskey(coin.platforms, "solana")
                address = coin.platforms.solana
                if !isempty(address) && !haskey(STATE.verified_tokens, address)
                    tokens[address] = Dict(
                        "name" => get(coin, "name", "Unknown"),
                        "symbol" => uppercase(get(coin, "symbol", "UNK")),
                        "type" => "token",
                        "verified" => true,
                        "official" => false,
                        "source" => source_name,
                        "coingecko_id" => get(coin, "id", "")
                    )
                    solana_count += 1
                end
            end
        end

        @info "Parsed $solana_count Solana tokens from $source_name"
    catch e
        @error "Error parsing $source_name" error=e
    end

    return tokens
end

# ========================================
# UTILITY FUNCTIONS
# ========================================

"""
Check if whitelist needs updating
"""
function _should_update()::Bool
    lock(STATE.lock) do
        if isnothing(STATE.last_update)
            return true
        end
        return (now(UTC) - STATE.last_update).value > (STATE.update_interval_seconds * 1000)
    end
end

"""
    add_to_whitelist!(address::String, info::Dict)

Manually add address to whitelist.
"""
function add_to_whitelist!(address::String, info::Dict)
    lock(STATE.lock) do
        STATE.verified_tokens[address] = merge(info, Dict(
            "verified" => true,
            "source" => "manual",
            "added_at" => now(UTC)
        ))
        @info "Added $address to whitelist manually"
    end
end

"""
    remove_from_whitelist!(address::String)

Remove address from whitelist.
"""
function remove_from_whitelist!(address::String)
    lock(STATE.lock) do
        if haskey(STATE.verified_tokens, address)
            delete!(STATE.verified_tokens, address)
            @info "Removed $address from whitelist"
        end
    end
end

end # module WhitelistService
